python/src/aac/lang/references.py

```diff
@@ -21,3 +22,198 @@ def filter_definitions_by_reference(definition_to_filter: Definition) -> bool:
+
+
+def _process_root(selector: str, selector_field: str, selector_value: str, definitions: list[Definition]) -> list[Definition]:
+    """Return definitions that match the root selector.  If no selector, return all definitions."""
+    keepers = []
+    if selector:
+        # handle root level selector
+        for definition in definitions:
+            if str(definition.get_top_level_fields()[selector_field]) == str(selector_value):
+                keepers.append(definition)
+    else:
+        # no root level selector, so just keep all definitions and move on
+        keepers.extend(definitions)
+    return keepers
+
+
+def _process_non_root(key: str, prefix: list[str], selector: str, selector_field: str, selector_value: str, definitions: list[Definition]) -> list[Definition]:
+    # initialize return value as empty string
+    keepers = []
+    for definition in definitions:
+        top_level_fields = definition.get_top_level_fields()
+        dict_list = _drill_into_nested_dict(prefix[1:], top_level_fields)
+
+        for def_dict in dict_list:
+            if key in def_dict.keys():
+                if selector:
+                    # since there is a selector, only keep the definition if it contains the specified selector field and value
+                    # note that selectors filter based on child field values
+                    value_for_key = def_dict[key]
+                    if type(value_for_key) is list:
+                        for item in value_for_key:
+                            if str(item[selector_field]) == str(selector_value):  # casting these seems like a hack
+
+                                keepers.append(definition)
+                                break
+                    else:
+                        if str(value_for_key[selector_field]) == str(selector_value):  # casting these seems like a hack
+                            keepers.append(definition)
+                else:
+                    # if there's no selector, keep the definition since it contains the segment key
+                    keepers.append(definition)
+    return keepers
+
+
+def _drill_into_nested_dict(search_keys: list[str], search_me: dict) -> list:
+    if len(search_keys) == 0:
+        return [search_me]
+
+    if search_keys[0] in search_me.keys():
+        next_level = search_me[search_keys[0]]
+        if isinstance(next_level, list):
+            items_to_return = []
+            # iterate through next level items and continue search
+            for item in next_level:
+                if isinstance(item, dict):
+                    i = _drill_into_nested_dict(search_keys[1:], item)
+                    if len(i) > 0:
+                        items_to_return.extend(i)
+            return items_to_return
+        elif isinstance(next_level, dict):
+            return _drill_into_nested_dict(search_keys[1:], next_level)
+        else:
+            return []
+    return []
```

Comment:
I was able to simplify this a touch by leveraging the dict function `get` which returns the value for the key or None, and that the result of the recursive function call will be a list.

```suggestion
    items_to_return = []
    if search_keys[0] in search_me.keys():
        next_level = search_me.get(search_keys[0])
        if isinstance(next_level, list):
            # iterate through next level items and continue search
            for item in next_level:
                items_to_return.extend(_drill_into_nested_dict(search_keys[1:], item))
        elif isinstance(next_level, dict):
            items_to_return = _drill_into_nested_dict(search_keys[1:], next_level)

    return items_to_return
```